### PR TITLE
Deploy to Vespa Cloud from python, talk to endpoints with mTLS

### PR DESCRIPTION
**Pull request by jonmv**
@thigm85

### CODE REVIEW

**Comment by jonmv** [comment]
I think I'm done here. Further python specific changes left to @thigm85, and review to be done by @oyving :)

File: python/vespa/vespa/application.py
Comment (jonmv) [inline python/vespa/vespa/application.py:93]: Not sure if this works against the local docker instance.

File: python/vespa/vespa/application.py
Comment (jonmv) [inline python/vespa/vespa/application.py:109]: Same here.

File: python/vespa/vespa/package.py
Comment (jonmv) [inline python/vespa/vespa/package.py:522]: This file can only be read by the user writing it, and is deleted immediately when closed. The `VespaCloud` closes this file on _its_ close, and is meant to be used like
```python
with VespaCloud('tenant', 'application', '/path/to/api_key') as cloud:
    ...
```

File: python/vespa/vespa/package.py
Comment (jonmv) [inline python/vespa/vespa/package.py:577]: Hard-coded container name here. Remember this is the container name changes (in `services.xml`).

File: python/vespa/vespa/package.py
Comment (jonmv) [inline python/vespa/vespa/package.py:643]: This deletes the data key/cert _forever_. Consider allowing the user to specify the certificate file, for persistent storage, if it's meaningful to allow the user to keep talking to its endpoint outside of the python code, or if the `VespaCloud` variable can't be kept in scope.

**Comment by jonmv** [comment]
Ok, so the lifecycles of the `VespaCloud` class and the actual deployment are a bit weird together. The actual deployment may live on (unless you `delete(...)` it), but the mTLS creds are lost with the `VespaCloud` object. At the same time, it makes sense to not `delete`, because deployment time from scratch is 15 minutes, but only 15 seconds if you already have a running instance. So, as long as all python code starts with deploying (which probably is a good idea anyway, to keep the instance from disappearing), and then keeps the `VespaCloud` object in scope for as long as it runs, this isn't a problem.

File: python/vespa/vespa/application.py
Comment (oyving) [inline python/vespa/vespa/application.py:26]: Please indicate the type of the string. Is it a filename, two PEM encoded objects, etc.

File: python/vespa/vespa/package.py
Comment (oyving) [reply python/vespa/vespa/package.py:522]: Is there a way to use an in-memory file object instead of a temporary on-disk file?

File: python/vespa/vespa/application.py
Comment (oyving) [inline python/vespa/vespa/application.py:26]: This should indicate the type in the `cert`. Is it a file name or the object itself?

**Review by oyving — CHANGES_REQUESTED** [review]

File: python/vespa/vespa/application.py
Comment (oyving) [reply python/vespa/vespa/application.py:26]: This is a duplicate of the comment above 🤷

File: python/vespa/vespa/package.py
Comment (jonmv) [reply python/vespa/vespa/package.py:522]: Absolutely not, unfortunately. It's been filed as a "bug" for 8 years, but nothing has happened.

Comment (jonmv) [reply python/vespa/vespa/package.py:522]: There were even made PRs to address the issue, but the maintainers rejected them because they wanted a "perfect" solution, but this solution has, of course, not arrived, nor does it seem like it will.

File: python/vespa/vespa/application.py
Comment (jonmv) [reply python/vespa/vespa/application.py:26]: Fixed.

Comment (jonmv) [reply python/vespa/vespa/application.py:26]: Yeah 🤷‍♂️

File: python/vespa/vespa/package.py
Comment (oyving) [reply python/vespa/vespa/package.py:522]: How pythonic

Comment (thigm85) [reply python/vespa/vespa/package.py:522]: Another option is to use 
```
vespa_cloud = VespaCloud("vespa-team", "ms-marco", "/Users/tmartins/sample_application/tmartins.vespa-team.pem")
app = vespa_cloud.deploy('from-notebook', app_package)
```
Then continue  doing all the operation like `app.feed_data_point(...)` and then use `app.close()` when you are done.

Does that makes sense @jonmv?

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:651]: I believe that using
```
vespa_cloud = VespaCloud("vespa-team", "ms-marco", "/Users/tmartins/sample_application/tmartins.vespa-team.pem")
app = vespa_cloud.deploy('from-notebook', app_package)
# run as many commands as one needs
app.close()
```
solves the case where the `VespaCloud` variable can't be kept in scope.

Comment (thigm85) [reply python/vespa/vespa/package.py:651]: I think it is best to let them generate the certificate themselves if they want to talk to the end-point outside of the python code. As the intent of this library is to make experiments from a python environment.

File: python/vespa/vespa/package.py
Comment (jonmv) [reply python/vespa/vespa/package.py]: Do you mean if it makes sense to move generation of data place key/cert to when you deploy, so it lives only while  a particular deployment (`Vespa` object) is in scope? I think that makes sense, but there's not much practical difference. You can avoid keeping the `VespaCloud` object in scope, but that's it. 
And for the certificates, it's possible to just never `.close()` the object, of course. But you'd have a hard time finding the certificate file without printing its name somewhere ... :)

Comment (jonmv) [reply python/vespa/vespa/package.py]:
```python
with VespaCloud(...) as cloud:
    with cloud.deploy(...) as app:
        # both variables are in scope now, as you can't have the Vespa in scope without the VespaCloud anyway
        # and the VespaCloud is intended to be closed regardless of where the certs are, due to its HTTPConnection
```

Comment (jonmv) [reply python/vespa/vespa/package.py]: I guess you could perhaps do 
```python
with VespaCloud(...) as cloud:
    app = cloud.deploy(...)

with app:
    # do stuff
```

Comment (jonmv) [reply python/vespa/vespa/package.py]: Anyway, these are all details. Do whatever you think is better :)

File: python/vespa/vespa/package.py
Comment (jonmv) [inline python/vespa/vespa/package.py:710]: Hehe, this is why I hate autoformatters :)

File: python/vespa/vespa/package.py
Comment (thigm85) [inline python/vespa/vespa/package.py:737]: @jonmv It would be nice to include the `deployment_message` argument when creating the `Vespa` instance. What would best represent the deployment message in the Vespa Cloud case?

**Review by thigm85 — CHANGES_REQUESTED** [review]

File: python/vespa/vespa/package.py
Comment (jonmv) [reply python/vespa/vespa/package.py:737]: I guess that depends. What is it used for? The whole log could be it. It's usually not that long for dev deployments. So, everything that's printed in `follow_deployment`.

File: python/vespa/vespa/application.py
Comment (jonmv) [reply python/vespa/vespa/application.py]: I initially believed it would be possible to send the actual cert and key, of course ... >_<

File: python/vespa/vespa/package.py
Comment (jonmv) [reply python/vespa/vespa/package.py:756]: Yes, you'd need to move the key/cert stuff to the `deploy(...)` call, but that's possible.

Comment (jonmv) [reply python/vespa/vespa/package.py:756]: You should prefer `with ...:`, though, to ensure resources are closed when exceptions are thrown.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:737]: Ok, for now it is just for informational purposes.

Comment (jonmv) [reply python/vespa/vespa/package.py:737]: The deployment message from the config server is in there, if you do substring search, etc..

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:756]: No, I don't need to change anything. My code above is wrong and I meant `vespa_cloud.close()` instead of `app.close()`

I don't need to move anything to `.deploy()` as the `Vespa` instance knows the location of the cert.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:710]: @oyving sold me on `black` a while ago, now it is hard to live without it.

Comment (jonmv) [reply python/vespa/vespa/package.py:710]: Yeah, pros and cons :)

File: python/vespa/vespa/application.py
Comment (thigm85) [reply python/vespa/vespa/application.py:94]: Yes, it works.

**Review by thigm85 — APPROVED** [review]

**Comment by thigm85** [comment]
@oyving Can you make a final review before we merge?

**Review by oyving — CHANGES_REQUESTED** [review]
There are some methods on the class that are very internal to the class itself and not part of the documented API. Would it make sense to prefix those with `_` (according to PEP-8) to make the intent of their internal-ness clear?

File: python/vespa/vespa/package.py
Comment (jonmv) [reply python/vespa/vespa/package.py:673]: No, it contains a message intended for the user.

Comment (oyving) [reply python/vespa/vespa/package.py:673]: Wouldn't it be better to return that value from the method instead? Library code printing stuff is a bit iffy. I know there's a contention here between regular library code and interactive notebooks, but it would be nice to see if there was a better way.

Comment (jonmv) [reply python/vespa/vespa/package.py:673]: Yes, moving it up one level is a good idea.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py]: `application_package_name` is unnecessary here as the info is available through `self.application_package.name`. Made the changes.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:672]: I don't think this was a left over. The idea is to print the deployment messages so that the user knows which stage he is. This `response["message"]` contains the following
```
Deployment started in run 6 of dev-aws-us-east-1c for vespa-team.ms-marco.from-notebook. This may take about 15 minutes the first time.
```
The entire message including the updates looks something like:
```
Deployment started in run 6 of dev-aws-us-east-1c for vespa-team.ms-marco.from-notebook. This may take about 15 minutes the first time.
INFO    [08:08:03]  Deploying platform version 7.276.8 and application version unknown ...
INFO    [08:08:04]  No services requiring restart.
INFO    [08:08:04]  Deployment successful.
INFO    [08:08:05]  Session 13618 for tenant 'vespa-team' prepared and activated.
INFO    [08:08:05]  ######## Details for all nodes ########
INFO    [08:08:05]  h711a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- container on port 4080 has config generation 13618, wanted is 13618
INFO    [08:08:05]  h712a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- logserver-container on port 4080 has config generation 13618, wanted is 13618
INFO    [08:08:05]  h713a.dev.aws-us-east-1c.vespa-external.aws.oath.cloud: expected to be UP
INFO    [08:08:05]  --- platform docker.ouroath.com:4443/vespa/centos-tenant:7.276.8
INFO    [08:08:05]  --- container-clustercontroller on port 19050 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- storagenode on port 19102 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- searchnode on port 19107 has config generation 13618, wanted is 13618
INFO    [08:08:05]  --- distributor on port 19111 has config generation 13618, wanted is 13618
INFO    [08:08:05]  Found endpoints:
INFO    [08:08:05]  - dev.aws-us-east-1c
INFO    [08:08:05]   |-- https://msmarco-container.from-notebook.ms-marco.vespa-team.aws-us-east-1c.dev.public.vespa.oath.cloud/ (cluster 'msmarco_container')
INFO    [08:08:06]  Installation succeeded!
```

Comment (thigm85) [reply python/vespa/vespa/package.py:672]: @oyving Would you change it?

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py:664]: This is a good question. It is not necessary but I thought it would be nice to have a copy in case the user wants to check the config files that got deployed.

A future change would include the option to deploy a user generated .zip file to Cloud from pyvespa, instead of building the .zip file from our limited application package API. 

That way assuming there will always be a .zip file in the `disk_folder` would tie those two use cases together. But for now it is mostly for the user convenience of checking the config files.

File: python/vespa/vespa/package.py
Comment (oyving) [reply python/vespa/vespa/package.py:672]: I think it should be changed, but it might be over-engineering it. It's your call. But as a minimum I would have a `self.output` file-handle that you write these messages to (that default to `sys.stdout`). That way the output can at least be controlled by the user if they wish.

**Comment by thigm85** [comment]
> There are some methods on the class that are very internal to the class itself and not part of the documented API. Would it make sense to prefix those with `_` (according to PEP-8) to make the intent of their internal-ness clear?

Yes, I agree. Made the changes for the classes involved in this PR.

File: python/vespa/vespa/package.py
Comment (oyving) [reply python/vespa/vespa/package.py:663]: I think that's a mixing of concerns. Also, wouldn't it be better for the user to just inspect the application package object and get the files printed to stdout?

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py]: Made the change. Check if it was that what you had in mind.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py]: Done.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py]: I removed the step that write zip file to disk. I can revisit that if needed.

File: python/vespa/vespa/package.py
Comment (thigm85) [reply python/vespa/vespa/package.py]: Done.

**Review by oyving — APPROVED** [review]

**Pull request merged**
